File: 1001-2000/1889-minimum-space-wasted-from-packaging/1889-minimum-space-wasted-from-packaging.cpp
const int MOD = 1e9+7;
class Solution {
public:
    int minWastedSpace(vector<int>& packages, vector<vector<int>>& boxes) {
        vector<long long> sum(packages.size() + 1);
        sum[0] = 0;
        for (int i = 0; i < packages.size(); ++i) {
            sum[i+1] = sum[i] + packages[i];
        }
        vector<int> count(1e5+5, 0);
        for (int p : packages) ++count[p];
        for (int i = 1; i < count.size(); ++i) count[i] += count[i-1];
        long long mn = 1e17;
        for (vector<int> &box : boxes) {
            sort(box.begin(), box.end());
            long long tot = 0;
            long long prev_cnt = 0;
            for (int b : box) {
                long long cnt = count[b];
                long long total_space = (cnt - prev_cnt) * b;
                long long used = sum[cnt] - sum[prev_cnt];
                prev_cnt = cnt;
                tot += total_space - used;
            }
            if (prev_cnt == packages.size()) {
                mn = min(mn, tot);
            }
        }
        if (mn == 1e17) {
            return -1;
        }
        return mn % MOD;
    }
};

File: 1001-2000/1889-minimum-space-wasted-from-packaging/1889-minimum-space-wasted-from-packaging.java
// class Solution {
//     public int minWastedSpace(int[] packages, int[][] boxes) {
        
//     }
// }
import java.util.Arrays;

class Solution {
    public int minWastedSpace(int[] packages, int[][] boxes) {
        int totalPackages = packages.length;
        int maxPackageSize = 0;
        for (int packageSize : packages) {
            if (packageSize > maxPackageSize) {
                maxPackageSize = packageSize;
            }
        }

        int[] packageCounts = new int[maxPackageSize + 1];
        long sumOfPackageSizes = 0;
        for (int packageSize : packages) {
            packageCounts[packageSize]++;
            sumOfPackageSizes += packageSize;
        }

        int[] prefixPackageCounts = new int[maxPackageSize + 1];
        for (int i = 1; i <= maxPackageSize; i++) {
            prefixPackageCounts[i] = prefixPackageCounts[i - 1] + packageCounts[i];
        }

        long minTotalBoxArea = Long.MAX_VALUE;

        for (int[] supplierBoxes : boxes) {
            Arrays.sort(supplierBoxes);
            
            if (supplierBoxes[supplierBoxes.length - 1] < maxPackageSize) {
                continue;
            }

            long currentSupplierBoxArea = 0;
            int lastPackedIndex = 0;

            for (int boxSize : supplierBoxes) {
                int currentPackedIndex = (boxSize >= maxPackageSize) ? totalPackages : prefixPackageCounts[boxSize];
                
                int packagesInCurrentBox = currentPackedIndex - lastPackedIndex;
                if (packagesInCurrentBox > 0) {
                    currentSupplierBoxArea += (long) packagesInCurrentBox * boxSize;
                }
                
                lastPackedIndex = currentPackedIndex;
                if (lastPackedIndex == totalPackages) {
                    break;
                }
            }
            
            if (currentSupplierBoxArea < minTotalBoxArea) {
                minTotalBoxArea = currentSupplierBoxArea;
            }
        }

        if (minTotalBoxArea == Long.MAX_VALUE) {
            return -1;
        }
        
        return (int) ((minTotalBoxArea - sumOfPackageSizes) % 1000000007);
    }
}